### ISSUE TITLE
New Failsafe Tab 

### ISSUE THREAD

**Issue by skaman82** [skaman82/cleanflight-configurator#45]

**Comment by skaman82** [comment]
Working with @ProDrone on concepts for this right now

**Comment by skaman82** [comment]
@ProDrone I made a template FailsafeTab for you to play with. Just check out my FailsafeTab branch.
Note1: Placed all text right in HTML in this stage. I can lace it in the language file afterwards.
Note2: I just copied the configuration.js file for the new template. You have to remove the unused functions.

![bildschirmfoto 2015-11-21 um 16 20 48](https://cloud.githubusercontent.com/assets/11768024/11319103/e29808f8-906b-11e5-81cf-95403e85e4ab.png)

**Comment by skaman82** [comment]
@ProDrone can you work with this?

**Comment by ProDrone** [comment]
@skaman82
Yes, sure, it looks good to me!

It will take me some time to connect the code. First i need to make changes to cleanflight code (changing the failsafe logic somewhat and add MSP messages to convey the info). Then i move out of my comfort zone and will add as much JS code as i can manage.

Also busy times with the day job :sweat_smile:

**Comment by ProDrone** [comment]
@skaman82 
I am working on this, starts looking good, but it's a fight and i learn a lot.

**Comment by ProDrone** [comment]
@skaman82 
__Update:__ I don't like the drop-down selector for AUX channels and the separation between flight channels and AUX channels. It is using less screen space the way it is now, but i prefer to have all channels listed like on the receiver tab.

It all works now apart from the flight channels. Still cleaning up to do and making the changes to generate a full channel list. I also moved fields around a bit.

**Comment by skaman82** [comment]
@ProDrone I also think it makes more sense to list all available channels. This way it is more accessible (fewer clicks). Can you list ONLY the channels that are available from the connected receiver? (If a user has a 6ch receiver it will only show 6 channels)

**Comment by ProDrone** [comment]
Yes, that is what i want to make, channels that are not in use are a waste of space and confusing too.

**Comment by ProDrone** [comment]
I completed the changes to the failsafe TAB. Merged your latest `new_stuff` branche to it, so that you are up to date to review it. The changes are in my repo at https://github.com/ProDrone/cleanflight-configurator/tree/FailsafeTab_ProDrone_mods

For testing the configurator changes you need cleanflight firmware at https://github.com/ProDrone/cleanflight/tree/add_msp_failsafe_support (needs modification to make it a proper pull request).

**Comment by ProDrone** [comment]
@skaman82 
I completed the changes to the failsafe TAB. Merged your latest new_stuff branche to it, so that you are up to date to review it. The changes are in my repo at https://github.com/ProDrone/cleanflight-configurator/tree/FailsafeTab_ProDrone_mods

For testing the configurator changes you need cleanflight firmware at https://github.com/ProDrone/cleanflight/tree/add_msp_failsafe_support (needs modification to make it a proper pull request).

@hydra Maybe you can take a look at this too?

**Comment by skaman82** [comment]
@ProDrone Great! I will have a look this evening. Could you provide a naze32 hex for testing? 
I don't know if merging the new_stuff branch was a a such good Idea in its actual state. The work on the GPS Tab is still not finished in that branch. @NightHawk32 is looking into it.

**Comment by ProDrone** [comment]
This is what it looks like:
![screenhunter_73](https://cloud.githubusercontent.com/assets/7206988/11461417/46f91fd8-9703-11e5-922c-26c1b954e265.jpg)
[cleanflight_NAZE.hex.pdf](https://github.com/skaman82/cleanflight-configurator/files/47046
/cleanflight_NAZE.hex.pdf)

Naze hex file above (remove the.pdf extension please).

The merging  was what i did to let you do a quick checkout from my fork for testing.
What would you like me to do? Open PR against your fork? Which branch?

**Comment by skaman82** [comment]
@ProDrone Tested this just now. Looking good! Great job with the tooltops - this helps a lot understanding the setup process. Three questions/suggestions:

1. There is still a failsafe box in the Config tab. I thought you wanted to remove it and make failsafe always on? We have now two "failsafe throttle" inputs in different tabs.

2. What is the difference between the flight channel settings "auto" und "hold". You should consider to clarify this in the tooltip. This is the one thing I still don't get.

3. Compatibility. Do you thought about compatibility to prior FC versions? I think there should at least be a message appearing when you use a older version of the firmware an all elements that are not compatible should be hidden. What do you think?

**Comment by ProDrone** [comment]
@skaman82 Thanks!
I consider it under construction until satisfied ;-) I am thinking of adding an indication what the AUX channels are assigned to. But  should it then replace the AUX n labels or put next to it? For consistency it should then also be on the receiver tab. What do you think?

1. The failsafe box and throttle setting on the config tab has to be conditionally hidden when the failsafe tab is completed (and in use based on firmware/MSP protocol version). I need to make modifications to cleanflight to keep failsafe always-on, add support for the drop quad and GPS procedures.

2. __Hold__ means the last received good pulse value is used for a channel. __Auto__ means that the flight channels are controlled by cleanflight (not directly following the user settings). Currently this means: Roll, Pitch and Yaw go to mid stick (`mid_rc`) and Throttle to minimum valid pulse value (`rx_min_usec`). I like to change that, but need to be very careful when making changes. Safety is easily compromised on quick changes. 

3. You are right. I was thinking of hiding the failsafe tab on older firmware/MSP protocol versions. But maybe it should be more fine grained, like you propose.

**Comment by skaman82** [comment]
Showing what the AUX channels are assigned to is a very great idea. I will think about how this could look like. Regarding 3. I don't think the unsupported fields shoud be grained, the should be not visible at all.

**Comment by ProDrone** [comment]
Up to now my changes are based on your branch `failsafeTab`, should i continue to build on top of that, and will you merge later? I will drop my current branch which is merged with you new stuff then, oke?

**Comment by skaman82** [comment]
yes, I think that would be better for now. 
We want to make a PR against the orig/dev version from the new_stuff branch in the next days.

**Comment by ProDrone** [comment]
I made some local mods. Still work in progress.

__These AUX channel mode assignments__
![screenhunter_74](https://cloud.githubusercontent.com/assets/7206988/11518281/481d09a4-9890-11e5-8055-b603814eaba5.jpg)

__Are now shown like this on the failsafe TAB__
![screenhunter_75](https://cloud.githubusercontent.com/assets/7206988/11518339/a2d127e0-9890-11e5-8de8-06d57ce2711a.jpg)

What do you think?

**Comment by ProDrone** [comment]
If you want to play with it, it is here: https://github.com/ProDrone/cleanflight-configurator/tree/failsafeTAB_ProDrone_v2

**Comment by skaman82** [comment]
It is starting to look very good! But you can have multiple modes on one aux switch like buzzer+althold+poshold. Here is not enough room for that and I would move the select fields at the last half of the box (Aux+Modes 60%; Select:25%; Input 15%). Also the modes could be styled differently than the chanel label. I will play with it a bit :)

**Comment by ProDrone** [comment]
Slightly changed, same link as above.

**Comment by ProDrone** [comment]
__Update on current failsafe TAB__
![screenhunter_76](https://cloud.githubusercontent.com/assets/7206988/11576734/32e2ab22-9a19-11e5-8fb2-85fafb9efff5.jpg)

[cleanflight_NAZE.hex.pdf](https://github.com/skaman82/cleanflight-configurator/files/51450/cleanflight_NAZE.hex.pdf)
Naze hex file above (remove the.pdf extension please).

The code for the configurator is at the same link as above.

**Comment by skaman82** [comment]
@proDrone Playing with this right now. Could you please place each Mode Name in a separate span with a class "modename"?

Von meinem iPhone gesendet

**Comment by ProDrone** [comment]
@skaman82 Do you mean this line?
https://github.com/ProDrone/cleanflight-configurator/blob/failsafeTAB_ProDrone_v2/tabs/failsafe.js#L115

And is that to style it in css, so that i maybe should remove all styling there?

**Comment by skaman82** [comment]
Yes I mean that line. At the moment all mode names are in this span. It would be better for styling if each name would be in a separate span. 

Von meinem iPhone gesendet

**Comment by ProDrone** [comment]
Do you mean that, when for instance 2 modes like ANGLE and BEEPER are assigned to AUX 2, that both modenames (ANGLE and BEEPER) need to be in their own span as opposed to being together in one span?

**Comment by skaman82** [comment]
exactly

**Comment by ProDrone** [comment]
*Just* updated accordingly. :grinning:

**Comment by skaman82** [comment]
Made small changes and send you the files via email.
Note: if you want to use green color, please use the brand color from the logo (its #59aa29). My version looks like this now:

<img width="925" alt="bildschirmfoto 2015-12-04 um 02 48 08" src="https://cloud.githubusercontent.com/assets/11768024/11579998/7a45eeb0-9a33-11e5-85bc-a36c68564306.png">

**Comment by ProDrone** [comment]
The files received by email are now included.

I made a modification to the style of the mode names because i did not like the light green color, it does not stand out that much (low contrast). For consistency i also removed the semicolons from the channel names (there are no other tabs that use this).

I experimented with more black/white like colors, but then it start to look like an input field. Next i tried the darker shade of green from the borders, but then it start to look like a button...

__I ended up with this:__
![screenhunter_77](https://cloud.githubusercontent.com/assets/7206988/11592026/740a51da-9a9c-11e5-8a9d-29b7f8677d02.jpg)
To me it now feels like an attached label, maybe you can come up with something better.

I want to round this part up now, but that requires coding to other, non-failsafe-tab parts.
The whole failsafe tab should be hidden when connected to older firmware version and the `receiver failsafe` section on the configuration tab need to be hidden when the failsafe tab is visible.

Do you want to take the current modifications in first and make these changes yourself, or should i change the non-failsafe-tab code and make it part of a my modifications? I am assuming you want to checkout my branch when i am finished, then merge it with your failsafeTab branch, and finally make a pull-request to the official cleanflight-configurator repo?

I already made changes to Cleanflight firmware code and want to continue with that ASAP so that both, the configurator and the firmware get in sync.

**Comment by skaman82** [comment]
OK, I will test some other colors. Now it looks like a note :) I also see some font problem in the SVGs in your screenshot. I will fix that too. I would always show the failsafe tab and display the content depending on firmware compatibility. 
It would be a bit complicated, don't you think? 
if you have firmware A you make your changes in tab A
if you have firmware B the the settings are in tab B

I suggest to move the current failsafe settings Box from the Config tab to the Failsafe tab and display only that settings box if the firmware is old. Otherwise hide it and display the new failsafe settings. So the settings will happen in the same place independent from firmware version.

I'dont need to merge this into my branch. You can make the PR directly. Its mostly your work after all and I don't need any credit :)

**Comment by ProDrone** [comment]
Hahaha, that is because i took the colors from the note. In the end it is both, a matter of taste and of clarity.

I see your point, but if you have firmware A, you are already used to where these settings are.
Then it does not change the current situation. It will only change when talking to newer firmware.

I like the idea of being consistent though, and that means putting it all on one new tab. I will go ahead and make it so...

**Comment by powdermnky007** [comment]
Thanks for the link skaman82.  Reposting my ideas here.  Btw this looks amazing!!!!  hats off to you both!  I slightly modified this from my posting on the other page as some of the ideas are now unnecessary.


1 Please include a drop down box with different presets. If not a a dropdown box that automatically sets them for you, you could put the information within a 'recommended settings example' button.  For example, one of the presets would be:

Disarm and Cut Throttle: 
Useful for miniquads. This preset will hold all stick values for a momentary loss of control, then on a full failsafe it will set throttle to zero and let it fall from the sky.

Upon selecting that option, these cli values would be set. I got these CLI values from ProDrone in the betaflight thread. 
rxfail 0 h
rxfail 1 h
rxfail 2 h
rxfail 3 h
rxfail 4 h
set failsafe_delay = 10
set failsafe_off_delay = 0
set failsafe_throttle = 1000


2 If a serial rx is enabled (Frsky X4r) it should have a warning telling user to set radio to 'No Pulses'. If CPPM/PWM is set, have a warning box that tells to consult their manual on how to do it. and give a big red scary thing saying FAILSAFE WILL NOT WORK UNLESS YOU DO THIS!!!!!

3 at the bottom have a write up on how to test failsafe. I know there are a few ways, but personally I go with the following method:
Remove Props
Plug battery into multirotor
Arm multirotor and throttle up until motors start spinning
Turn radio off
If motors stop and multirotor starts beeping you are done!
Note: Multirotor will not rearm until power its power is cycled

**Comment by skaman82** [comment]
@powdermnky007  regarding 1. I would say stage 2 setting (drop) does exactly what you want

**Comment by skaman82** [comment]
@ProDrone what about this? I think this looks good and is very readable.

``
.tab-failsafe .modename {
    background-color: #828885;
    border-radius: 3px;
    border: 1px solid #686c6a;
    color: #fff !important;
    font-weight:normal;
    font-family: 'open_sanssemibold', Arial;
    padding-left: 3px;
    padding-right:3px;
    margin-right:3px;
}
´´

**Comment by powdermnky007** [comment]
regarding 1. I would say stage 2 setting (drop) does exactly what you want.

@skaman82    Perfect, thank you

**Comment by skaman82** [comment]
@ProDrone this just occurred to me: Are you also making sure that all failsafe settings are going in to the backup file generated by GUI backup function?

**Comment by ProDrone** [comment]
@skaman82 Thanks. I did not (yet?) think of that.

**Comment by ProDrone** [comment]
@skaman82 I did some more work on the failsafe tab (changes pushed, so you may check it out).
Support for both, the old (like in current configuration tab) and the new style is in the failsafe tab.
This means that this may be removed from the configuration tab when new failsafe tab get in.

Still have to look at the changes in your pull-request.

**Comment by skaman82** [comment]
@ProDrone you can safely merge my pr - it is just some simple css stuff and removal of the old failsafe box (just hidden). Will try your new version later today.

**Comment by skaman82** [comment]
@ProDrone just tested this quick. Really great work with all the tooltips an explanations! 

I noticed that I am unable to save Failsafe Throttle in the new tab  - it always jumps back to 1000. This only happens in the new style where this input is in stage2 settings - old style box works fine inside the failsafe tab.

I'm guessing you didn't include the settings into the backup functionality of the GUI yet. Restoring the settings from the backup file didn't work.

BTW. You should update you configurator branch and include the newest changes from the dev version (in you version there is still a bug in the firmware flasher tab)

**Comment by ProDrone** [comment]
@skaman82
* Merged the development branch in.
* Corrected the save `failsafe_throttle` value error.
* Cherry picked your `Small Styling changes`.
* Changes pushed.

I did not modify the backup/restore yet, first try to focus on completing the failsafe tab.

**Comment by skaman82** [comment]
@ProDrone Looking good! Just tested. Which tasks are still open regarding the new tab?

**Comment by ProDrone** [comment]
@skaman82 
1. All message texts need to go into messages.json
2. TESTING :grinning: 

I am quite happy with the current state...

**Comment by skaman82** [comment]
quick question about failsafe config. Trying configure failsafe for the first time. In my case if I turn my RX off, the throttle and AUX channel values are about 1060. If I turn RX on they go to 1000. How the system can detect a signal loss if the channel values are still valid? I want to configure a landing procedure but when I turn my RX off the motors just stop. Do I setup something wrong?

**Comment by ProDrone** [comment]
The system can only detect the failsafe state by looking at the received values and in case of a serial receiver by checking a failsafe indicator flag (if that is supported by the particular receiver).

Your receiver should be configured to produce out of range values on failsafe when it is not a serial type with failsafe state flag. Some have an option to send no-pulses on failsafe.

I tried to cover several cases, like:
1. FC is powered on without a receiver connected.
2. FC is powered with receiver ON and the TX is still off (should never be done like that, but it happens..)
3. FC is powered on with receiver ON and TX ON. TX goes OFF after some time.

What RX/TX do you use?

Did you read the https://github.com/cleanflight/cleanflight/blob/master/docs/Rx.md ?

**Comment by skaman82** [comment]
I am using different orange RX (R616XN, R615X, R620X) with DX6i. The only difference I have with 620X is the beeper goes off when quad is on and TX is off. Silence with others. Do you know how to configure the RX to send no pulses?

**Comment by ProDrone** [comment]
I am not familiar with Spektrum but i think it does not have the no-pulses configuration option.

You receiver should be set so that all channels have a range from 1000-2000, with 1500 being the middle position.

What you can do after that is, using the trim function, move one of your flightchannels (throttle is good) as low as it can go. Then lock in this value as the failsafe value for your receiver (that should be in the manual i hope).

Finally set the minimum valid pulse length in the middle between 1000 and the lowest value you achieved, so that this is detected on receiver failsafe state.

Hope this makes sense...

**Comment by powdermnky007** [comment]
That is why I suggested earlier to have some instructions/guide for 'normal users'.  BTW you two have done a great job.

**Comment by le3dmax** [comment]
@skaman82 
You need to set the failsafe on your Orange receiver. Set the min throttle on your Radio below 1000 and rebind it. It tells the receiver to drop below 1000 when you lose the connection.
Once you did that you can back you radio to 1000 again.

**Comment by skaman82** [comment]
thank you all, this worked perfectly and yes, @powdermnky007 is right. This should be at least in the documentation .

**Comment by powdermnky007** [comment]
Yeah I watched someones brand new miniquad fly away, never to be seen again.  That prompted my friends and I to double check our failsafe settings.  They were all setup incorrectly.  Every time my friends build a new quad, they believe they have their failsafe set correctly, but it never is.  I always help them set it correctly.  Some of them say, "yeah... I don't understand all that 'computer stuff' I just want to fly."    makes me sigh :\

Again, thank you for your hard work here.  Hopefully this will save many quadcopters in the future :)

**Comment by ProDrone** [comment]
@powdermnky007 It is still work in progress, but i adapted the manual here: https://github.com/ProDrone/cleanflight/blob/add_msp_failsafe_support/docs/Failsafe.md

And for the case you mentioned above i did this today: https://github.com/ProDrone/cleanflight/blob/f79bdc94bb9928ef38a31187ab90c7d17e29f8c8/src/main/rx/rx.c#L404-L407

:grinning:

**Comment by skaman82** [comment]
Just tested all the Failsafe Tabs settings with R616XN and 620X. Works wonderful :)

**Comment by powdermnky007** [comment]
@ProDrone just read manual and it looked great and very thorough.  I like how the very 1st step in configuration is to set receiver to no pulses... good placement :)  if that is overlooked all can be lost!

As far as the code, I don't fully understand it, but it half was makes sense to me.  But I don't need to understand it... heh I trust you :)  good job.

That is the first time i've seen the 'failsafe_procedure' cli variable.  Not sure exactly how that fits in, as you can set it to do that with the following:
set failsafe_delay = 10
set failsafe_off_delay = 0
set failsafe_throttle = 1000
You may want to elaborate on how that works?  if it manipulates other cli variables?  I'm sure there will be questions about that in forums.  But i don't get it.  The values drop / land are simple enough though.

**Comment by skaman82** [comment]
The variable failsafe_procedure can be used for other things in the future like RTH or just the RTH landing procedure when they integrated - see digitalentitys iNAV (@ProDrone correct me if I am wrong)

**Comment by skaman82** [comment]
@ProDrone you should also include the situation a was in about a hour ago in your manual to help with exactly the same problem. Something like this:

"If you have a RX that is not configured to send no pulses on signal loss you can configure failsafe by doing following steps: 
1. Adjust you throttle channel with trims to a minimum possible value (throttle should go way below 1000)
2. Bind with your TX in this state. 
3. Restore the trim to a neutral value
4. Switch you TX off and check the channel value in the configurator (throttle should now go way below 1000)
5. Set the minimum puls range setting in the failsafe tab somewhere between you lowest RX value and 1000"

**Comment by ProDrone** [comment]
@powdermnky007
Setting to no pulses does not apply to many RX's. There are so many different setups out there to cover. It is important to make it clear that a receiver _somehow_ needs to inform the flight-controller about it's state.

The code change will make the quad __not__ fly away when stage 2 is disabled. As it currently is people can set it wrong and then...well what you witnessed. I like to see the possibility to disable this removed in the near future.

The variables `failsafe_off_delay` and `failsafe_throttle` are used by the only failsafe procedure that is in place since the beginning, which is __auto-land__. More procedures will be added, like __drop__ which is new and intended for racing quads. I expect that most procedures will have their own set of variables to tinker with.

@skaman82 
I get your point. Good instructions are valuable. But i'm not sure if that should be added to the failsafe manual. Maybe to the RX manual? We should avoid repeating what is already in the manual that comes with your RX/TX.

And now... back to the code...

**Comment by skaman82** [comment]
@ProDrone Just checked the manuals. There is a lot of information how to adjust values in the bind process that are activated on failsafe. Orange RX come without a manual or just 1 copied page. 
The CRITICAL "TRICK" with the TRIMS to send a invalid pulse is nowhere to be found. It might sound logical now but that didn't occurred to me before. The Failsafe manual ist the place where I would personally look first to get all the settings right. I Would never place this in the RX manual because the instructions are specially failsafe related.

**Comment by ProDrone** [comment]
@skaman82 I agree that starting to read the failsafe manual is logical, but from there you may be directed to the receiver documentation (rx.md) and from there to the manuals for specific equipment.

Strictly speaking correct, but on the other hand: it would not hurt to have step by step setup guides for popular setups too.

I looked into the mods you made to configuration.css / .htm / .js but don't understand how you made the old failsafe settings disappear. I like to get rid of all code in there instead of hiding it, so that it may become part of the same pull-request.

__Can you maybe checkout my latest code and make that correction for me?__

**Comment by skaman82** [comment]
@ProDrone I have hidden the old box just to make the css adjustments to align the other boxes after the failsafe box would be deleted (just to simulate it to make proper stylig). I thought you will just delete the box afterwards. I gladly can remove it completely in HTML but I have no idea what to remove in the JS file. 

Also I can help you placing the Text in the json file if you want to.

**Comment by ProDrone** [comment]
@skaman82
Thanks, all texts are already done. Works good.
Just attempted the remove the failsafe part from configuration tab, but that is wrong now...

**Comment by ProDrone** [comment]
@skaman82 I think everything is in place now. Can you please review it?

**Comment by skaman82** [comment]
@prodrone still at work :(, will do later in the evening and give you a feedback

**Comment by ProDrone** [comment]
This is my cleanfligh test-firmware version (.pdf appended for GitHub acceptance):
[cleanflight_NAZE.hex.pdf](https://github.com/skaman82/cleanflight-configurator/files/57254/cleanflight_NAZE.hex.pdf)

Not a big difference with the previous (added forcing auto-mode to flight-channels when stage 2 is disabled).

**Comment by ProDrone** [comment]
@skaman82 I have to admit that i never looked at the backup/restore buttons in the setup tab.
I think a backup utility that warns you that it will make an incomplete image is useless.
To be sure that ALL is conserved i always used the CLI dump command.

I cannot imagine that 'normal users' feel confident with this. How are they suppose to know if that precious setting is a CLI setting? __*All*__ settings are CLI settings, some happen to be manipulated by the GUI.

**Comment by skaman82** [comment]
@ProDrone NightHawk32 and I want to change things regarding that. Using cli for backup is just wrong. We will try to make the backup function save all settings completely but this is serious heavy work. If you can include your failsafe settings now it would be a tiny bit less to do for us.

**Comment by ProDrone** [comment]
Oké, i understand. I will have a look...

**Comment by skaman82** [comment]
Thanks! This issue tracker is our ToDo list for things we want to improve. If you can/want help out with something of that youre welcome :)

https://github.com/skaman82/cleanflight-configurator/issues

**Comment by ProDrone** [comment]
@skaman82 The failsafe tab variables are now added to Backup & Restore.
I did put a note on line 520 of file `backup_restore.js` as a reminder (and for feedback).

**Comment by skaman82** [comment]
@ProDrone thanks, just tested the backup function. Works like a charm! :) Thank you

**Comment by skaman82** [comment]
Tested the Failsafe functionality again, no problems.

**Comment by ProDrone** [comment]
What do you think about the point i mentioned on line 520 of file `backup_restore.js` ?

**Comment by skaman82** [comment]
@NightHawk32 could you please answer ProDrones question? I believe you added this.

**Comment by NightHawk32** [comment]
Dominic did this, but I see your point @ProDrone there we only add redundant data which is already in the firmware. I will have a look at that and kick out this part and do a reset instead, as you suggested.

**Comment by ProDrone** [comment]
@NightHawk32
__Thanks !__
I already did this to test it out. My knowledge of JS is limited and i am still figuring things out. What i see is that, when restoring an older (api) backup, the newer additions to the configuration are becoming undefined, like when in this function:
~~~C
function load_objects() {
  MISC = configuration.MISC;
  RC_MAP = configuration.RCMAP;
  BF_CONFIG = configuration.BF_CONFIG;
  SERIAL_CONFIG = configuration.SERIAL_CONFIG;
  LED_STRIP = configuration.LED_STRIP;
  ARMING_CONFIG = configuration.ARMING_CONFIG;
  FC_CONFIG = configuration.FC_CONFIG;
  _3D = configuration._3D;
  RX_CONFIG = configuration.RX_CONFIG;
  FAILSAFE_CONFIG = configuration.FAILSAFE_CONFIG;
  RXFAIL_CONFIG = configuration.RXFAIL_CONFIG;
}
~~~
The undefined `configuration.RX_CONFIG` is assigned to the newer configuration object `RX_CONFIG`, so it becomes also undefined. Then later in the code the MSP messages cause exceptions.

I think it should work more robust. The latest configuration is always already defined in the data storage and should not be destroyed because of a data dependency. It should just be filled with the data from a backup-file. But maybe i overlook things (certainly do ;-) and the current code may be brilliant...

Still digging.

**Comment by ProDrone** [comment]
@NightHawk32 @skaman82 
I decided to go with the flow and made my modifications in line with the existing code.
(previous comment deleted for clarity). Now it all works as intended, also when restoring older backups.

Beware, i did a forced push (rewriting history), so checkout a fresh copy, here:
https://github.com/ProDrone/cleanflight-configurator/tree/failsafeTAB_ProDrone_v2

**Comment by ProDrone** [comment]
@skaman82 Updated again: stage 2 elements reorganized. When stage 2 is disabled they are all hidden. Moved the settings relevant for landing inside the land-pane as a visual clue. These field are only enabled when land procedure is selected. Adapted help-texts and some more...

What do you think?

**Comment by ProDrone** [comment]
Looks like this now:
![screenhunter_78](https://cloud.githubusercontent.com/assets/7206988/11748687/bb442796-a029-11e5-9683-2d037dbf047c.jpg)

**Comment by skaman82** [comment]
@ProDrone Great Idea. This makes a lot of sense to rearrange the fields this way.

**Comment by ProDrone** [comment]
@skaman82 The GUI side is almost done. However, before making a pull-request, the required changes to cleanflight need to be merged first, because without these changes this will not work.

I have more plans to make this even better, but think this should be merged first, as there currently is no good configuration method. I hope @hydra can find the time to take a look at this, since there are things i like to change in cleanflight that needs his approval.

**Comment by ProDrone** [comment]
@skaman82 So i wrapped it up and made a pull request for both, cleanflight and configurator. Hoping it will soon be merged to fill the need to configure failsafe via GUI.

**Comment by skaman82** [comment]
Great news, we are about to finish our next set of improvements for the configurator. The changes we made also require a api bump. 

Von meinem iPhone gesendet

**Comment by ProDrone** [comment]
@skaman82 Do you mean that there also is (or will be) a pull-request in cleanflight that is required for the for your work?

**Comment by skaman82** [comment]
Yes, there will be a PR shortly 

Von meinem iPhone gesendet

**Comment by skaman82** [comment]
Done. Closing this. Thanks @ProDrone - this was fun! Would like to work with you again on other things.
